src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -175,4 +159,184 @@ export class MagicWandService {
+    // imgData.data (DataArray).
+    const indexAsDataArray: number =
+        this.coordToDataArrayIndex(xCoord, yCoord, imgData.width);
+    visited.add(indexAsDataArray);
+
+    // Performs a basic floodfill if the scribbles set is empty.
+    if (scribbles.size === 0) {
+      // Loops until no more adjacent pixels within tolerance level.
+      while (visit.length !== 0) {
+        const coord: Array<number> = visit.pop();
+        // Unpacks coord
+        const x: number = coord[0];
+        const y: number = coord[1];
+
+        // Operational part of while-loop.
+        mask.add(this.coordToDataArrayIndex(x, y, imgData.width));
+
+        // Loop part of while-loop.
+
+        // Gets coords of adjacent pixels.
+        const neighbors: Array<Array<number>> =
+            [[x - 1, y], [x + 1, y], [x, y - 1], [x, y + 1]];
+        // Adds coords of adjacent pixels to the heap.
+        for (const neighborPixel of neighbors) {
+          const x: number = neighborPixel[0];
+          const y: number = neighborPixel[1];
+          // Checks if coord is in bounds and has not been visited first.
+          if (!this.getIsValid(imgData.width, imgData.height, x, y, visited)) {
+            continue;
+          }
+          // Visits the pixel and checks if it should be part of the mask.
+          visited.add(this.coordToDataArrayIndex(x, y, imgData.width));
+          if (this.getIsMask(originalPixel, imgData, neighborPixel, tolerance)) {
+            visit.push(neighborPixel);
+          }
+        }
+      }  // End of while loop.
+    } else {
+      // Performs the scribble floodfill if the scribbles set is not empty.
+
+      // Loops until no more adjacent pixels within tolerance level.
+      while (visit.length !== 0) {
+        const coord: Array<number> = visit.pop();
+        // Unpacks coord.
+        const x: number = coord[0];
+        const y: number = coord[1];
+
+        // Operational part of while-loop.
+        mask.add(this.coordToDataArrayIndex(x, y, imgData.width));
+
+        // Loop part of while-loop.
+
+        // Gets coords of adjacent pixels.
+        const neighbors: Array<Array<number>> =
+            [[x - 1, y], [x + 1, y], [x, y - 1], [x, y + 1]];
+        // Adds coords of adjacent pixels to the heap.
+        for (const neighborPixel of neighbors) {
+          const x: number = neighborPixel[0];
+          const y: number = neighborPixel[1];
+          // Checks if coord is in bounds and has not been visited first.
+          if (!this.getIsValid(imgData.width, imgData.height, x, y, visited)) {
+            continue;
+          }
+          // Visits the pixel and check if it should be part of the mask.
+          visited.add(this.coordToDataArrayIndex(x, y, imgData.width));
+          if (this.getIsScribbleMask(scribbles, imgData, neighborPixel, tolerance)) {
+            visit.push(neighborPixel);
+          }
+        }
+      }  // End of while loop.
+    }
```

Comment:
We are getting close.  If you notice the if/else code is exactly the same code except for:

`if (this.getIsMask(originalPixel, imgData, neighborPixel, tolerance))`

and 

`if (this.getIsScribbleMask(scribbles, imgData, neighborPixel, tolerance))`

So you could do two things:

Remove outer if/else and add it to lines 300.  So it would be:

```
if (scribbles.size === 0) {
  if (this.getIsMask(originalPixel, imgData, neighborPixel, tolerance)) {
    visit.push(neighborPixel);
   }
} else {
  if (this.getIsScribbleMask(scribbles, imgData, neighborPixel, tolerance)) {
    visit.push(neighborPixel);
  }
}
```

I think you might be able to do even better.  

One option is to create a helper function:

```
getIsMask(originalPixel, scribbles,  imgData, neighborPixel, tolerance) {
  if (scribbles.size === 0) {
    if (this.getIsMask(originalPixel, imgData, neighborPixel, tolerance)) {
      visit.push(neighborPixel);
     }
  } else {
    if (this.getIsScribbleMask(scribbles, imgData, neighborPixel, tolerance)) {
      visit.push(neighborPixel);
    }
  }
}
```

This is better because then the floodfill code is easier to read, but can we do better?

If I understand it correctly I think we can get the same functionality out of getIsMask by using getIsScribbleMask. If we call `this.getIsScribbleMask(new Set<number>(originalPixel), imgData, neighborPixel, tolerance)` then it will return the same thing.

So I would recommend that you change floodFill to:

```
floodfill(imgData: ImageData, xCoord: number, yCoord: number,
      tolerance: number): Set<number> {
    scribbles = new Set<number>(this.pixelIndexToXYCoordToPixelIndex(xCoord, yCoord))
    return this.doFloodfill(imgData, xCoord, yCoord, tolerance, scribbles);
  }
```